libraries/AR_Motors/AP_MotorsUGV.cpp

```diff
@@ -213,8 +220,99 @@ void AP_MotorsUGV::set_throttle(float throttle)
         return;
     }
 
+    
     // check throttle is between -_throttle_max and  +_throttle_max
-    _throttle = constrain_float(throttle, -_throttle_max, _throttle_max);
+    float throttle_chk = constrain_float(throttle, -_throttle_max, _throttle_max);
+    
+    if (_esc_backward_unlock == 0)
+    {
+        _throttle = throttle_chk;
+        return ;
+    }
+
+    /*
+    ** ((throttle_chk == 0) && (_esc_backward_unlock_active)) continue unlocking if user releases throttle but it is zero not forward.
+    ** (throttle_chk < 0)  backward mode anyway
+    */
+    if ((throttle_chk < 0.0f) || ((throttle_chk <= 0.0f) && (_esc_backward_unlock_active) ))
+    {
+        const uint32_t now = AP_HAL::millis();
+        
+        switch (_esc_backward_unlock_state)
+        {
+            case UnlockState::UNLOCK_WAITING:
+                _esc_backward_unlock_simulated_signal=0;
+                _throttle = 0.0f;
+                _last_set_throttle_call = now;
+                _esc_backward_unlock_active = true;
+                _esc_backward_unlock_state = UnlockState::UNLOCK_STEP_DEC_TO_MAX_NEG;
+                break;
+
+            case UnlockState::UNLOCK_STEP_DEC_TO_MAX_NEG:
+                if ((now - _last_set_throttle_call) >= 20)
+                {
+                    _esc_backward_unlock_simulated_signal-=2.0f;
+                    _throttle = _esc_backward_unlock_simulated_signal;
+                }
+                _last_set_throttle_call = now;
```

Comment:
Doesn't this get stuck forever if the recall time is less than 20ms? Looks like your just comparing this calls`now` against last calls `now`. If the loop is fast enough you will never change `_esc_backward_unlock_simulated_signal` and so never reach the threshold to move on to the next case.

Reply:
Ooopps, thank you... this is a bug .... I reversed the logic to get out if less than 20ms
I also didnt added brakets for break I found this is the  norm in all the code in ardu.